### PR TITLE
T6477: Add telegraf loki output plugin

### PR DESCRIPTION
**Pull request by sever-sever**
## Change Summary

Add Loki plugin to telegraf

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

 * https://vyos.dev/T6477

## Component(s) name

telegraf
## How to test

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

**Review by dmbaturin — APPROVED** [review]

**Comment by sever-sever** [comment]
@MergifyIo backport circinus sagitta

File: interface-definitions/service_monitoring_telegraf.xml.in [hunk 1]
```diff
@@ -148,6 +148,36 @@
...
+                      #include <include/generic-password.xml.i>
+                    </children>
+                  </node>
+                  <leafNode name="metric-name-label">
+                    <properties>
+                      <help>Metric name label</help>
+                      <valueHelp>
+                        <format>txt</format>
+                        <description>Label to use for the metric name</description>
+                      </valueHelp>
```
Comment (c-po) [inline interface-definitions/service_monitoring_telegraf.xml.in:171]: Please add a constraint like `constraint/alpha-numeric-hyphen-underscore-dot.xml.i`

File: src/conf_mode/service_monitoring_telegraf.py [hunk 2]
```diff
@@ -159,6 +162,11 @@ def verify(monitoring):
         if 'url' not in monitoring['splunk']:
             raise ConfigError(f'Monitoring splunk "url" is mandatory!')
 
+    # Verify Loki
+    if 'loki' in monitoring:
+        if 'url' not in monitoring['loki']:
+            raise ConfigError(f'Monitoring loki "url" is mandatory!')
```
Comment (c-po) [inline src/conf_mode/service_monitoring_telegraf.py:168]: `url` should be uppercase

File: src/conf_mode/service_monitoring_telegraf.py [hunk 3]
```diff
@@ -159,6 +162,11 @@ def verify(monitoring):
         if 'url' not in monitoring['splunk']:
             raise ConfigError(f'Monitoring splunk "url" is mandatory!')
```
Comment (c-po) [inline src/conf_mode/service_monitoring_telegraf.py:163]: `url` should be uppercase

**Review by c-po — CHANGES_REQUESTED** [review]

**Comment by c-po** [comment]
Should there probably a `verify()` check that if username is specified, so must be password and vice versa?

File: src/conf_mode/service_monitoring_telegraf.py (hunk 3 above)
Comment (sever-sever) [reply src/conf_mode/service_monitoring_telegraf.py:163]: The `url` is the missed option of the configuration, not the URL
Do we need to change it to the URL here?
```
vyos@r4# set service monitoring telegraf splunk 
Possible completions:
 > authentication       HTTP basic authentication parameters
   url                  Remote URL

```

File: src/conf_mode/service_monitoring_telegraf.py (hunk 2 above)
Comment (sever-sever) [reply src/conf_mode/service_monitoring_telegraf.py:168]: The `url` is the missed option of the config, not the URL
Do we need URL here?

File: interface-definitions/service_monitoring_telegraf.xml.in (hunk 1 above)
Comment (sever-sever) [reply interface-definitions/service_monitoring_telegraf.xml.in:171]: Done!

**Comment by sever-sever** [comment]
> Should there probably a `verify()` check that if username is specified, so must be password and vice versa?

Done!

File: src/conf_mode/service_monitoring_telegraf.py (hunk 2 above)
Comment (c-po) [reply src/conf_mode/service_monitoring_telegraf.py:168]: Yes, as it's a abbreviation. `raise ConfigError('Monitoring loki URL is mandatory!')`

File: src/conf_mode/service_monitoring_telegraf.py (hunk 3 above)
Comment (c-po) [reply src/conf_mode/service_monitoring_telegraf.py:163]: No, I only mean the URL portion of the error message should be uppercase:

`raise ConfigError('Monitoring splunk URL is mandatory!')`

File: src/conf_mode/service_monitoring_telegraf.py (hunk 2 above)
Comment (sever-sever) [reply src/conf_mode/service_monitoring_telegraf.py:168]: Done!

File: src/conf_mode/service_monitoring_telegraf.py (hunk 3 above)
Comment (sever-sever) [reply src/conf_mode/service_monitoring_telegraf.py:163]: Done!

File: smoketest/scripts/cli/test_service_monitoring_telegraf.py [hunk 4]
```diff
@@ -63,5 +63,23 @@ def test_01_basic_config(self):
         for input in inputs:
             self.assertIn(input, config)
 
+    def test_02_loki(self):
+        label = 'r123'
+        loki_url = 'http://localhost'
+        port = '3100'
+
+        self.cli_set(base_path + ['loki', 'url', loki_url])
+        self.cli_set(base_path + ['loki', 'port', port])
```
Comment (c-po) [inline smoketest/scripts/cli/test_service_monitoring_telegraf.py:75]: Please add test code for username and password and also check if the exception is present when only username is specified but no password

**Review by c-po — CHANGES_REQUESTED** [review]

File: smoketest/scripts/cli/test_service_monitoring_telegraf.py (hunk 4 above)
Comment (sever-sever) [reply smoketest/scripts/cli/test_service_monitoring_telegraf.py:75]: Done!

**Review by c-po — APPROVED** [review]

**Pull request merged**